plugins/amqp/src/plugin.cpp

```diff
@@ -584,9 +607,10 @@ class rabbitmq_saver final : public plugin_saver {
 
   auto instantiate(operator_control_plane& ctrl, std::optional<printer_info>)
     -> caf::expected<std::function<void(chunk_ptr)>> override {
-    auto engine = amqp_engine::make(config_);
-    if (not engine)
-      return engine.error();
+    auto _engine = amqp_engine::make(config_);
+    if (not _engine)
+      return _engine.error();
+    auto engine = std::make_shared<amqp_engine>(std::move(*_engine));
```

Comment:
```suggestion
    auto engine = std::make_shared<amqp_engine>{};
    if (auto eng = amqp_engine::make(config_)) {
       engine.reset(std::move(*eng));
    } else {
       return eng.error();
    } 
```

auto engine = std::make_shared<amqp_engine>(std::move(*_engine));

Reply:
Yep, looks cleaner. Thanks!

Reply:
Hmm...
```
[ 92%] Building CXX object plugins/amqp/CMakeFiles/amqp-shared.dir/src/plugin.cpp.o
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp: In member function ‘virtual caf::expected<std::function<void(caf::intrusive_ptr<tenzir::chunk>)> > tenzir::plugins::amqp::{anonymous}::rabbitmq_saver::instantiate(tenzir::operator_control_plane&, std::optional<tenzir::printer_info>)’:
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:637:48: error: expected ‘,’ or ‘;’ before ‘{’ token
  637 |     auto engine = std::make_shared<amqp_engine>{};
      |                                                ^
```
If I change the first line to:
```c++
            auto engine = std::make_shared<amqp_engine>();
```
I get
```
[ 92%] Building CXX object plugins/amqp/CMakeFiles/amqp-shared.dir/src/plugin.cpp.o
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp: In member function ‘virtual caf::expected<std::function<void(caf::intrusive_ptr<tenzir::chunk>)> > tenzir::plugins::amqp::{anonymous}::rabbitmq_saver::instantiate(tenzir::operator_control_plane&, std::optional<tenzir::printer_info>)’:
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:639:19: error: no matching function for call to ‘std::shared_ptr<tenzir::plugins::amqp::{anonymous}::amqp_engine>::reset(std::remove_reference<tenzir::plugins::amqp::{anonymous}::amqp_engine&>::type)’
  639 |       engine.reset(std::move(*eng));
      |       ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/12/bits/shared_ptr.h:53,
                 from /usr/include/c++/12/memory:77,
                 from /home/satta/tmp/tenzir/libtenzir/aux/caf/libcaf_core/caf/fwd.hpp:8,
                 from /home/satta/tmp/tenzir/libtenzir/include/tenzir/fwd.hpp:16,
                 from /home/satta/tmp/tenzir/libtenzir/include/tenzir/concept/convertible/data.hpp:11,
                 from /home/satta/tmp/tenzir/libtenzir/include/tenzir/argument_parser.hpp:11,
                 from /home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:9:
/usr/include/c++/12/bits/shared_ptr_base.h:1646:9: note: candidate: ‘template<class _Yp> std::__shared_ptr<_Tp, _Lp>::_SafeConv<_Yp> std::__shared_ptr<_Tp, _Lp>::reset(_Yp*) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
 1646 |         reset(_Yp* __p) // _Yp must be complete.
      |         ^~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1646:9: note:   template argument deduction/substitution failed:
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:639:19: note:   mismatched types ‘_Yp*’ and ‘tenzir::plugins::amqp::{anonymous}::amqp_engine’
  639 |       engine.reset(std::move(*eng));
      |       ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1655:9: note: candidate: ‘template<class _Yp, class _Deleter> std::__shared_ptr<_Tp, _Lp>::_SafeConv<_Yp> std::__shared_ptr<_Tp, _Lp>::reset(_Yp*, _Deleter) [with _Deleter = _Yp; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
 1655 |         reset(_Yp* __p, _Deleter __d)
      |         ^~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1655:9: note:   template argument deduction/substitution failed:
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:639:19: note:   mismatched types ‘_Yp*’ and ‘tenzir::plugins::amqp::{anonymous}::amqp_engine’
  639 |       engine.reset(std::move(*eng));
      |       ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1660:9: note: candidate: ‘template<class _Yp, class _Deleter, class _Alloc> std::__shared_ptr<_Tp, _Lp>::_SafeConv<_Yp> std::__shared_ptr<_Tp, _Lp>::reset(_Yp*, _Deleter, _Alloc) [with _Deleter = _Yp; _Alloc = _Deleter; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
 1660 |         reset(_Yp* __p, _Deleter __d, _Alloc __a)
      |         ^~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1660:9: note:   template argument deduction/substitution failed:
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:639:19: note:   mismatched types ‘_Yp*’ and ‘tenzir::plugins::amqp::{anonymous}::amqp_engine’
  639 |       engine.reset(std::move(*eng));
      |       ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1641:7: note: candidate: ‘void std::__shared_ptr<_Tp, _Lp>::reset() [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
 1641 |       reset() noexcept
      |       ^~~~~
/usr/include/c++/12/bits/shared_ptr_base.h:1641:7: note:   candidate expects 0 arguments, 1 provided
In file included from /usr/include/c++/12/optional:44,
                 from /home/satta/tmp/tenzir/build/libtenzir/include/tenzir/config.hpp:4,
                 from /home/satta/tmp/tenzir/libtenzir/include/tenzir/fwd.hpp:11:
/usr/include/c++/12/bits/stl_construct.h: In instantiation of ‘constexpr void std::_Construct(_Tp*, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}]’:
/usr/include/c++/12/bits/alloc_traits.h:635:19:   required from ‘static constexpr void std::allocator_traits<std::allocator<void> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}; allocator_type = std::allocator<void>]’
/usr/include/c++/12/bits/shared_ptr_base.h:604:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr_base.h:971:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Alloc = std::allocator<void>; _Args = {}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr_base.h:1712:14:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr.h:464:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine]’
/usr/include/c++/12/bits/shared_ptr.h:1009:14:   required from ‘std::shared_ptr<std::_NonArray<_Tp> > std::make_shared(_Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}; _NonArray<_Tp> = tenzir::plugins::amqp::{anonymous}::amqp_engine]’
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:637:48:   required from here
/usr/include/c++/12/bits/stl_construct.h:115:28: error: no matching function for call to ‘construct_at(tenzir::plugins::amqp::{anonymous}::amqp_engine*&)’
  115 |           std::construct_at(__p, std::forward<_Args>(__args)...);
      |           ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/12/bits/stl_construct.h:94:5: note: candidate: ‘template<class _Tp, class ... _Args> constexpr decltype (::new(void*(0)) _Tp) std::construct_at(_Tp*, _Args&& ...)’
   94 |     construct_at(_Tp* __location, _Args&&... __args)
      |     ^~~~~~~~~~~~
/usr/include/c++/12/bits/stl_construct.h:94:5: note:   template argument deduction/substitution failed:
/usr/include/c++/12/bits/stl_construct.h: In substitution of ‘template<class _Tp, class ... _Args> constexpr decltype (::new(void*(0)) _Tp) std::construct_at(_Tp*, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}]’:
/usr/include/c++/12/bits/stl_construct.h:115:21:   required from ‘constexpr void std::_Construct(_Tp*, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}]’
/usr/include/c++/12/bits/alloc_traits.h:635:19:   required from ‘static constexpr void std::allocator_traits<std::allocator<void> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}; allocator_type = std::allocator<void>]’
/usr/include/c++/12/bits/shared_ptr_base.h:604:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr_base.h:971:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Alloc = std::allocator<void>; _Args = {}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr_base.h:1712:14:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr.h:464:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine]’
/usr/include/c++/12/bits/shared_ptr.h:1009:14:   required from ‘std::shared_ptr<std::_NonArray<_Tp> > std::make_shared(_Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}; _NonArray<_Tp> = tenzir::plugins::amqp::{anonymous}::amqp_engine]’
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:637:48:   required from here
/usr/include/c++/12/bits/stl_construct.h:96:17: error: no matching function for call to ‘tenzir::plugins::amqp::{anonymous}::amqp_engine::amqp_engine()’
   96 |     -> decltype(::new((void*)0) _Tp(std::declval<_Args>()...))
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:262:3: note: candidate: ‘tenzir::plugins::amqp::{anonymous}::amqp_engine::amqp_engine(tenzir::plugins::amqp::{anonymous}::amqp_engine&&)’
  262 |   amqp_engine(amqp_engine&& other) noexcept
      |   ^~~~~~~~~~~
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:262:3: note:   candidate expects 1 argument, 0 provided
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:218:12: note: candidate: ‘tenzir::plugins::amqp::{anonymous}::amqp_engine::amqp_engine(tenzir::plugins::amqp::{anonymous}::amqp_config)’
  218 |   explicit amqp_engine(amqp_config config)
      |            ^~~~~~~~~~~
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:218:12: note:   candidate expects 1 argument, 0 provided
/usr/include/c++/12/bits/stl_construct.h: In instantiation of ‘constexpr void std::_Construct(_Tp*, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}]’:
/usr/include/c++/12/bits/alloc_traits.h:635:19:   required from ‘static constexpr void std::allocator_traits<std::allocator<void> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}; allocator_type = std::allocator<void>]’
/usr/include/c++/12/bits/shared_ptr_base.h:604:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr_base.h:971:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Alloc = std::allocator<void>; _Args = {}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr_base.h:1712:14:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/12/bits/shared_ptr.h:464:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {}; _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine]’
/usr/include/c++/12/bits/shared_ptr.h:1009:14:   required from ‘std::shared_ptr<std::_NonArray<_Tp> > std::make_shared(_Args&& ...) [with _Tp = tenzir::plugins::amqp::{anonymous}::amqp_engine; _Args = {}; _NonArray<_Tp> = tenzir::plugins::amqp::{anonymous}::amqp_engine]’
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:637:48:   required from here
/usr/include/c++/12/bits/stl_construct.h:119:7: error: no matching function for call to ‘tenzir::plugins::amqp::{anonymous}::amqp_engine::amqp_engine()’
  119 |       ::new((void*)__p) _Tp(std::forward<_Args>(__args)...);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:262:3: note: candidate: ‘tenzir::plugins::amqp::{anonymous}::amqp_engine::amqp_engine(tenzir::plugins::amqp::{anonymous}::amqp_engine&&)’
  262 |   amqp_engine(amqp_engine&& other) noexcept
      |   ^~~~~~~~~~~
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:262:3: note:   candidate expects 1 argument, 0 provided
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:218:12: note: candidate: ‘tenzir::plugins::amqp::{anonymous}::amqp_engine::amqp_engine(tenzir::plugins::amqp::{anonymous}::amqp_config)’
  218 |   explicit amqp_engine(amqp_config config)
      |            ^~~~~~~~~~~
/home/satta/tmp/tenzir/plugins/amqp/src/plugin.cpp:218:12: note:   candidate expects 1 argument, 0 provided
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning’ may have been intended to silence earlier diagnostics
make[2]: *** [plugins/amqp/CMakeFiles/amqp-shared.dir/build.make:76: plugins/amqp/CMakeFiles/amqp-shared.dir/src/plugin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:3829: plugins/amqp/CMakeFiles/amqp-shared.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

Reply:
Sorry, that should have been:

```cpp
            auto engine = std::shared_ptr<amqp_engine>{};
```

Reply:
Should probably be
~~~
auto engine = std::shared_ptr<amqp_engine>{};
~~~

Reply:
Thanks!